### PR TITLE
Extend README with preliminary contribution notes

### PR DESCRIPTION
**Pull request by nelson-lojo**

### CODE REVIEW

File: src/function_classes/linear.py [hunk 1]
```diff
@@ -33,7 +33,7 @@ def __init__(self, sparsity: int = 3, *args, **kwargs):
         self._sparsity = sparsity
 
     def evaluate(self, x_batch: torch.Tensor, *params: torch.Tensor) -> torch.Tensor:
-        weights, _ = params
+        weights, *_ = params
```
Comment (riensou) [inline src/function_classes/linear.py:36]: the correct fix here is `weights = torch.stack(params)` I believe, but this is fixed anyway in #31

File: README.md [hunk 2]
```diff
@@ -18,3 +18,45 @@ Resume a training run specified by `<config_file>` starting from `<run_id>/<chec
...
+
+
+## Extending this repo
+
+We follow the following flow for objects in this repository:
+![image of the flow map](<readme_assets/flow map.png>)
+
+These initialized classes are passed the dictionary parsed directly from yaml as arguments. We recommend reading examples for your respective function class in [conf/](https://github.com/in-context-learning-2024/in-context/blob/main/conf/)
+
+Make sure to initialize your parent class with `super().__init__(...)` in both cases
```
Comment (riensou) [inline README.md:30]: to me, reads a bit unclear. 'These initialized classes' and 'in both cases' could maybe be reworded to be more precise. (i.e. which classes and which cases?)

File: README.md [hunk 3]
```diff
@@ -18,3 +18,45 @@ Resume a training run specified by `<config_file>` starting from `<run_id>/<chec
...
+
+In addition, we require both `.context_length: int` and `.name: str` to be set for evaluation and visualization.
+
+---
+### Adding Function Classes
+
+We take the approach of framing each function class as an iterator, where the boilerplate of making the iterator work for our training scheme (`ContextTrainer`) is [implemented for you](https://github.com/in-context-learning-2024/in-context/blob/main/src/core/function_class.py). The order of calls against a `FunctionClass` is as follows:
+- We `__init__` the `FunctionClass` with the parsed yaml configuration
+    - this calls `_init_param_dist()`, which should produce a [`torch.distributions.Distribution`](https://pytorch.org/docs/stable/distributions.html#distribution) 
+        - Note: If the implemented ones in pytorch are not enough, we also provide a way to ["combine" distributions](https://github.com/in-context-learning-2024/in-context/blob/main/src/utils.py)
```
Comment (riensou) [inline README.md:54]: change to "Note: If the distributions implemented in PyTorch are not enough, ..."

File: README.md [hunk 4]
```diff
@@ -18,3 +18,45 @@ Resume a training run specified by `<config_file>` starting from `<run_id>/<chec
...
+
+---
+### Adding Models
+
+We wrap all sequence architectures and baselines as [`ContextModel`s](https://github.com/in-context-learning-2024/in-context/blob/main/src/core/context_model.py). The order of calls against a `ContextModel` is as follows:
+- We `__init__` the `ContextModel` directly from the yaml configuration
+- Inside the training loop and during evaluation, we call `forward` to generate predictions
+
+`ContextModel`s thus have only two required methods:
+- `__init__(...)`, and
```
Comment (riensou) [inline README.md:43]: remove the ", and" maybe to be consistent with the list of required methods in the next section

File: README.md [hunk 5]
```diff
@@ -18,3 +18,45 @@ Resume a training run specified by `<config_file>` starting from `<run_id>/<chec
...
+
+These initialized classes are passed the dictionary parsed directly from yaml as arguments. We recommend reading examples for your respective function class in [conf/](https://github.com/in-context-learning-2024/in-context/blob/main/conf/)
+
+Make sure to initialize your parent class with `super().__init__(...)` in both cases
+
+We try to make it easy to contribute by [adding models](#adding-models) or [function classes](#adding-function-classes):
+
+
+---
+### Adding Models
```
Comment (riensou) [inline README.md:36]: should we add an explanation of how to get yaml files to 'see' their added context model? like, adding it to the \_\_init\_\_.py file and then adding it to models.yml?

File: README.md [hunk 6]
```diff
@@ -18,3 +18,45 @@ Resume a training run specified by `<config_file>` starting from `<run_id>/<chec
...
+- Inside the training loop and during evaluation, we call `forward` to generate predictions
+
+`ContextModel`s thus have only two required methods:
+- `__init__(...)`, and
+- `forward(self, xs: torch.Tensor, ys: torch.Tensor) -> torch.Tensor`
+
+In addition, we require both `.context_length: int` and `.name: str` to be set for evaluation and visualization.
+
+---
+### Adding Function Classes
```
Comment (riensou) [inline README.md:49]: should we add an explanation of how to get yaml files to 'see' their added function class? like, adding it to the \_\_init\_\_.py file?

**Review by riensou — CHANGES_REQUESTED** [review]
overall looks very good so far, some minor edits suggested

File: src/function_classes/linear.py (hunk 1 above)
Comment (nelson-lojo) [reply src/function_classes/linear.py:36]: I think this is right, review that PR

File: README.md (hunk 4 above)
Comment (nelson-lojo) [reply README.md:43]: makes sense. done

File: README.md (hunk 2 above)
Comment (nelson-lojo) [reply README.md:30]: resolved by saying `FunctionClass` and `ContextModel`

File: README.md (hunk 3 above)
Comment (nelson-lojo) [reply README.md:54]: done.

File: README.md (hunk 5 above)
Comment (nelson-lojo) [reply README.md:36]: this is a good point. added

File: README.md (hunk 6 above)
Comment (nelson-lojo) [reply README.md:49]: added.

**Review by riensou — APPROVED** [review]
all looks good now

**Pull request merged**